tools/reviewer/cas_proxy/Proxy.java

```diff
@@ -0,0 +1,200 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.startupos.tools.reviewer.cas_proxy;
+
+import com.google.common.hash.Hashing;
+import com.google.protobuf.ByteString;
+import com.google.startupos.common.firestore.FirestoreProtoClient;
+import com.google.startupos.tools.reviewer.proxy.AuthProtos.FileDownloadRequest;
+import com.google.startupos.tools.reviewer.proxy.AuthProtos.FileDownloadResponse;
+import com.google.startupos.tools.reviewer.proxy.AuthProtos.FileUploadRequest;
+import com.google.startupos.tools.reviewer.proxy.AuthProtos.FileUploadResponse;
+import com.google.startupos.tools.reviewer.proxy.CASProxyServiceGrpc;
+
+import io.grpc.Server;
+import io.grpc.ServerBuilder;
+import io.grpc.stub.StreamObserver;
+
+import java.io.ByteArrayOutputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public class Proxy {
+
+  // TODO(vmax): make bucket configurable
+  public static final String CAS_BUCKET_NAME = "test-bucket-cas";
+
+  static class AccessDeniedException extends Exception {}
+
+  static class Service extends CASProxyServiceGrpc.CASProxyServiceImplBase {
+    private FirestoreProtoClient client;
+    private AccessManager accessManager;
+    private HashMap<String, String> fileCache = new HashMap<>();
+
+    static class FileUploader implements StreamObserver<FileUploadRequest> {
+
+      private StreamObserver<FileUploadResponse> responseObserver;
+      private FirestoreProtoClient protoClient;
+      private ByteArrayOutputStream buffer;
+      private AccessManager accessManager;
+      private AtomicBoolean hasAccess = new AtomicBoolean(false);
+
+      public FileUploader(
+          FirestoreProtoClient protoClient,
+          AccessManager accessManager,
+          StreamObserver<FileUploadResponse> responseObserver) {
+        this.responseObserver = responseObserver;
+        this.protoClient = protoClient;
+        this.buffer = new ByteArrayOutputStream();
+        this.accessManager = accessManager;
+      }
+
+      @Override
+      public void onNext(FileUploadRequest value) {
+        if (this.accessManager.hasWriteAccess(value.getIdToken())) {
+          hasAccess.set(true);
+        }
+
+        if (!hasAccess.get()) {
+          responseObserver.onError(new AccessDeniedException());
+        }
+
+        try {
+          buffer.write(value.getContent().toByteArray());
+        } catch (IOException e) {
+          e.printStackTrace();
+          responseObserver.onError(e);
+        }
+      }
+
+      @Override
+      public void onError(Throwable t) {
+        responseObserver.onError(t);
+      }
+
+      @Override
+      public void onCompleted() {
+        String sha256 = Hashing.sha256().hashBytes(buffer.toByteArray()).toString();
```

Comment:
So we keep everything in the buffer, in memory?
Why not write it to file on `onNext()` and then close it on `onCompleted()`?

Reply:
@oferb Right! I'll add a `TODO` and handle in new iteration

Reply:
@oferb `todo` in https://github.com/google/startup-os/pull/541/commits/d751e511a2d7fe83561daa67f717aa702dc3da71